### PR TITLE
EIP4844 Chain Updater

### PR DESCRIPTION
**Pull request by mehdi-aouadi**
## PR Description
Add blobs management to the Chain Upter

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/util/DataStructureUtil.java [hunk 1]
```diff
@@ -1962,4 +1983,21 @@ public Spec getSpec() {
   private <T> T getConstant(final Function<SpecConfig, T> getter) {
     return getter.apply(spec.getGenesisSpec().getConfig());
   }
+
+  private Bytes getSampleBlob() {
```
Comment (zilm13) [inline ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/util/DataStructureUtil.java]: I'd make this one public and call something like `randomCorrectBlob` in oppose to `randomBlob`, or we could measure it and maybe leave only correct one with name `randomBlob` if it's not significantly slow.
And also if we make it public here, we could use it in `KZGTest` and remove its copy from it

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java [hunk 2]
```diff
@@ -417,11 +439,22 @@ private void assertBlockCanBeGenerated() {
...
+        .getSpecMilestoneAtSlot(slot)
+        .isGreaterThanOrEqualTo(SpecMilestone.EIP4844);
+  }
+
   private void trackBlock(final SignedBlockAndState block) {
     blocks.put(block.getSlot(), block);
     blocksByHash.put(block.getRoot(), block);
   }
 
+  private void trackBobsSidecar(final BlobsSidecar blobsSidecar) {
```
Comment (zilm13) [inline ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java]: trackB`l`obsSidecar

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java [hunk 3]
```diff
@@ -63,44 +70,51 @@
...
 import tech.pegasys.teku.spec.signatures.LocalSigner;
 import tech.pegasys.teku.spec.signatures.Signer;
+import tech.pegasys.teku.spec.util.DataStructureUtil;
 
 /** A utility for building small, valid chains of blocks with states for testing */
 public class ChainBuilder {
   private static final List<BLSKeyPair> DEFAULT_VALIDATOR_KEYS =
       Collections.unmodifiableList(new MockStartValidatorKeyPairFactory().generateKeyPairs(0, 3));
-
+  private static final int RANDOM_BLOBS_COUNT = 10;
```
Comment (zilm13) [inline ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java]: We have 4 blobs maximum in spec for mainnet, so I'd keep this number low too, 4 or maybe even 2

**Review by zilm13 — COMMENTED** [review]
Could we also add any simple single test to this PR to be sure it's really working. I don't know, blobs are pruned on finalized fork check or something else.

**Comment by tbenr** [comment]
I built `BlobsUtil` class on top of Dmitry's #6566. I used it for the EL stub. Once we merge that I'll showup that PR so we can integrate that with the ChainUpdater if it fits well (hopefully it does)

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/util/DataStructureUtil.java (hunk 1 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/util/DataStructureUtil.java]: It's not possible to use the `DataStructureUtil` in the KZG module (Dependencies hierarchy issue).
We decided to keep the KZG tests blobs generation logic separated and created another one (less dry and more spec related).

Comment (zilm13) [reply ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/util/DataStructureUtil.java]: Yeah, I'm sleeping, it's text fixtures. Sorry, we cannot use it in kzg
anyway I'd make this one public and change name to be clear within random generators

**Comment by mehdi-aouadi** [comment]
@zilm13 FYI I reused the `BlobsUtil` to generate the blobs and removed it from the `DataStructureUtil` as discussed

File: ethereum/spec/build.gradle [hunk 4]
```diff
@@ -47,6 +47,7 @@ dependencies {
     testFixturesImplementation 'org.hyperledger.besu:evm'
     testFixturesImplementation 'org.hyperledger.besu:plugin-api'
     testFixturesImplementation 'org.junit.jupiter:junit-jupiter-api'
+    testFixturesImplementation 'tech.pegasys:jc-kzg-4844'
```
Comment (tbenr) [inline ethereum/spec/build.gradle]: you don't need this anymore

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java [hunk 5]
```diff
@@ -441,24 +473,101 @@ private SignedBlockAndState appendNewBlockToChain(final UInt64 slot, final Block
...
-                  options.getEth1Data(),
-                  options.getTransactions(),
-                  options.getTerminalBlockHash(),
-                  options.getExecutionPayload(),
-                  options.getBlsToExecutionChange(),
-                  options.getKzgCommitments(),
-                  options.getSkipStateTransition()));
+
+      if (eip4844MilestoneReached(slot) && options.getGenerateRandomBlobs()) {
+        List<Blob> randomBlobs = blobsUtil.generateBlobs(slot, RANDOM_BLOBS_COUNT);
```
Comment (tbenr) [inline ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java]: I'd move the code from each branch to a private method so it's easier to read

**Review by tbenr — APPROVED** [review]
LGTM with a couple things to improve

File: ethereum/spec/build.gradle (hunk 4 above)
Comment (mehdi-aouadi) [reply ethereum/spec/build.gradle]: good catch, done

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java (hunk 5 above)
Comment (mehdi-aouadi) [reply ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java]: done

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/BlockProposalTestUtil.java [hunk 6]
```diff
@@ -321,6 +328,77 @@ public SafeFuture<SignedBlockAndState> createBlock(
...
+      final Optional<Bytes32> terminalBlock,
+      final Optional<ExecutionPayload> executionPayload,
+      final Optional<SszList<SignedBlsToExecutionChange>> blsToExecutionChange,
+      final List<Blob> blobs,
+      final boolean skipStateTransition)
+      throws EpochProcessingException, SlotProcessingException {
+    final UInt64 newEpoch = spec.computeEpochAtSlot(newSlot);
+
+    final MiscHelpersEip4844 miscHelpers =
+        (MiscHelpersEip4844) spec.forMilestone(SpecMilestone.EIP4844).miscHelpers();
```
Comment (zilm13) [inline ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/BlockProposalTestUtil.java:351]: How about `spec.forMilestone(SpecMilestone.EIP4844).miscHelpers().toVersionEip4844().orElseThrow()`?

File: ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java [hunk 7]
```diff
@@ -441,31 +473,145 @@ private SignedBlockAndState appendNewBlockToChain(final UInt64 slot, final Block
...
+                options.getExecutionPayload(),
+                options.getBlsToExecutionChange(),
+                randomBlobs,
+                options.getSkipStateTransition()));
+
+    final BlobsSidecarSchema blobsSidecarSchema =
+        spec.getGenesisSchemaDefinitions().toVersionEip4844().orElseThrow().getBlobsSidecarSchema();
+
+    final MiscHelpersEip4844 miscHelpers =
+        (MiscHelpersEip4844) spec.forMilestone(SpecMilestone.EIP4844).miscHelpers();
```
Comment (zilm13) [inline ethereum/spec/src/testFixtures/java/tech/pegasys/teku/spec/generator/ChainBuilder.java:602]: How about `spec.forMilestone(SpecMilestone.EIP4844).miscHelpers().toVersionEip4844().orElseThrow()`?

**Review by zilm13 — APPROVED** [review]
LGTM

**Pull request merged**